File: scripts/CheckVariableMulticollinearity.py
# -*- coding: utf-8 -*-
# ---------------------------------------------------------------------------
# Check variable multicollinearity
# Author: Timm Nawrocki, Alaska Center for Conservation Science
# Created on: 2017-01-06
# Usage: Must be executed as an ArcPy Script.
# Description: This tool provides a check for multicollinearity in a set of variables per a given presence-absence dataset based on qr matrix decomposition.
# ---------------------------------------------------------------------------

# Import arcpy module
import arcpy, os, subprocess

# Set overwrite option
arcpy.env.overwriteOutput = True

# Define folder of environmental raster datasets
Raster_Folder = arcpy.GetParameterAsText(0)

# Define input species presence/absence points
Species_Points = arcpy.GetParameterAsText(1)

# Define output raster
Output_Text = arcpy.GetParameterAsText(2)

# Define location of RScript.exe
RScript = arcpy.GetParameterAsText(3)

# Define location of R Script
R_Script = arcpy.GetParameterAsText(4)

# Setup workspace
Workspace, Output_Name = os.path.split(Output_Text)

# Define intermediate files for user selected workspace
Species_Shapefile = os.path.join(Workspace, "Species_Points.shp")
Arguments = os.path.join(Workspace, "Arguments.txt")

# Copy species points to workspace as a shapefile
arcpy.CopyFeatures_management(Species_Points, Species_Shapefile)

# Prepare arguments
arcpy.AddMessage("Compiling R script...")
Raster_Folder_Path = str(Raster_Folder)
Species_Shapefile_Path = str(Species_Shapefile)
Output_Text_Path = str(Output_Text)
Workspace_Path = str(Workspace)
Raster_Folder_Path = "'" + Raster_Folder_Path + "'"
Species_Shapefile_Path = "'" + Species_Shapefile_Path + "'"
Output_Text_Path = "'" + Output_Text_Path + "'"

# Write arguments to text file
text_file = open(Arguments, "w")
text_file.write(Raster_Folder_Path + " " + Species_Shapefile_Path + " " + Output_Text_Path)
text_file.close()

# Send process to R
arcpy.AddMessage("Starting R...")
cmd = RScript + " " + R_Script + " " + Workspace
arcpy.AddMessage(cmd)
subprocess.call(cmd)

# Rename R Text to Output Text
R_Text = os.path.join(Workspace, "R_Text.txt")
R_Text_Open = open(R_Text, "r")
R_Text_Lines = R_Text_Open.readlines()
R_Text_Open.close()
Output_Text_Open = open(Output_Text, "w")
Output_Text_Open.write(str(R_Text_Lines))
Output_Text_Open.close()

# Delete intermediate data
arcpy.Delete_management(Species_Shapefile)
os.remove(Arguments)
os.remove(R_Text)
DirPath = os.path.split(R_Script)[0]
Info = os.path.join(DirPath, "info")
ArcDir = os.path.join(Info, "arc.dir")
if os.path.exists(ArcDir) == True:
    os.remove(ArcDir)
    os.rmdir(Info)
R_History = os.path.join(Workspace, ".Rhistory")
if os.path.exists(R_History) == True:
    os.remove(R_History)
XML = os.path.join(Output_Text, "txt.xml")
if os.path.exists(XML) == True:
    os.remove(XML)

File: scripts/CheckVariableMulticollinearity.R
# -*- coding: utf-8 -*-
# ---------------------------------------------------------------------------
# Generate random forest distribution raster
# Author: Timm Nawrocki, Alaska Center for Conservation Science
# Created on: 2016-11-21
# Usage: Must be executed as an R Script.
# Description: This tool provides a simple Random Forests model.
# ---------------------------------------------------------------------------

# Install required libraries if they are not already installed.
Required_Packages <- c("randomForest", "rfUtilities", "sp", "raster", "rgdal", "stringr", "ROCR")
New_Packages <- Required_Packages[!(Required_Packages %in% installed.packages()[,"Package"])]
if (length(New_Packages) > 0) {
  install.packages(New_Packages)
}

# Import required libraries: randomForest, rfUtilities, sp, raster, rgdal, stringr, ROCR.
library(randomForest)
library(rfUtilities)
library(sp)
library(raster)
library(rgdal)
library(stringr)
library(ROCR)

# Set seed
set.seed(415)

# Set workspace folder from command line input.
Workspace <- commandArgs(trailingOnly = TRUE)
setwd(Workspace)

# Import arguments passed by python.
Arguments <- read.table("Arguments.txt", header = FALSE)

# Set path to raster folder and generate raster stack from arguments passed by python.
Raster_Folder <- as.vector(Arguments[,1])
cat("Creating raster stack...\n")
Raster_List <- list.files(Raster_Folder, pattern="img$", full.names = TRUE)
Raster_Stack <- stack(Raster_List)

# Select species data from arguments passed by python.
Species_Shapefile <- as.vector(Arguments[,2])

# Define output raster from arguments passed by python.
Output_Text <- as.vector(Arguments[,3])

# Convert shapefile into spatial data frame and extract raster data.
cat("Creating species data frame...\n")
Species_Data <- readOGR(dsn = Species_Shapefile)
Species_Data@data <- data.frame(Species_Data@data, extract(Raster_Stack, Species_Data))

# Test for multiple colinearity in environmental variables with p value assigned based on number of variables.
cat("Performing colinearity tests...\n")
Raster_Number <- as.numeric(length(Raster_List))
if (Raster_Number < 20) {
  p_value <- 0.0000001
} else if (Raster_Number >= 20) {
  p_value <- 0.05
}
Collinearity <- multi.collinear(Species_Data@data, p = p_value)

# Remove variables one at a time to ensure that each variable identified as colinear is a true multiple colinearity.
if (length(Collinearity) != 0) {
  Collinear_List <- c(Collinearity)
  for (l in Collinearity) {
    Collinearity_Iteration <- multi.collinear(Species_Data@data[,-which(names(Species_Data@data)==l)], p = p_value)
    Collinear_List <- c(Collinear_List, Collinearity_Iteration)
  }
  # Generate list of variables to remove because of true multiple colinearity.
  Collinear_Remove <- character()
  for (l in Collinearity) {
    Word_Count <- as.numeric(sum(str_count(Collinear_List, l)))
    if (Word_Count == as.numeric(length(Collinearity))) {
      Collinear_Remove <- c(Collinear_Remove, l)
    }
  }
} else if (length(Collinearity) == 0) {
  Collinear_Remove <- "No multicollinear variables identified"
}

# Output list of collinear variables to text
sink("R_Text.txt")
cat(as.vector(Collinear_Remove))
sink()